File: pdcsl-scripts/src/main/resources/scripts/Plugins/PDCSL/OC_Volume_Counter.py
# OC_Volume_Counter - Compute volume occupied by OncoChrome fluorophores
# Copyright © 2019 Damien Goutte-Gattat

from ij import IJ
from qmul.pdcsl.helper import extract_volumes
from qmul.pdcsl.channelop import apply_channel_op


def run_script():
    img = IJ.getImage()

    ops = [
            [2, 'MASK(Moments)'],
            [1, 'MASK(Moments)'],
            [4, 'MASK(Moments)'],
            [3, 'MASK(MaxEntropy)']
            ]
    masks = apply_channel_op(img, ops)
    masks.show()

    values = extract_volumes(masks, range(4))

    IJ.log("{},{:.2f},{:.2f},{:.2f},{:.2f}".format(img.getTitle(), *values))


run_script()

File: src/main/java/uk/ac/qmul/bci/pdcsl/imagej/OCVolumeCounter.java
/*
 * PDCSL-ImageJ - PDCS Lab’s ImageJ Collection
 * Copyright © 2019 Damien Goutte-Gattat
 */

package uk.ac.qmul.bci.pdcsl.imagej;

import org.incenp.imagej.ChannelMasker;
import org.incenp.imagej.Helper;
import org.scijava.command.Command;
import org.scijava.plugin.Parameter;
import org.scijava.plugin.Plugin;
import org.scijava.ui.UIService;

import ij.ImagePlus;
import ij.measure.ResultsTable;

@Plugin(type = Command.class, menuPath = "Plugins>PDCSL>OC Volume Counter")
public class OCVolumeCounter implements Command {

	private static ChannelMasker masker = ChannelMasker.createMasker("G:MASK(Huang),C:MASK(Moments),G:MASK(Moments),Y:MASK(Moments),R:MASK(MaxEntropy)");

    @Parameter
    private ImagePlus image;

    @Parameter(label = "Order of channels:")
    private String channelOrder;

    @Parameter
    private UIService uiService;

    @Override
    public void run() {
        ImagePlus masks = masker.apply(image, image.getTitle() + " Masks", channelOrder);
        uiService.show(masks);

        ResultsTable rt = Helper.getResultsTable("OC Volume Counter");
        rt.incrementCounter();
        rt.addLabel(image.getTitle());

        double[] volumes = Helper.extractVolumes(masks);
        for ( int i = 0; i < volumes.length; i++ ) {
            rt.addValue(String.format("Channel %d", i + 1), volumes[i]);
        }
        rt.show("OC Volume Counter");
    }

}
